File: leetcode/easy/155-minStack.java
// solution 1
class MinStack {
    private ArrayList<Integer> stack;
    private ArrayList<Integer> min;
    
    public MinStack() {
        stack = new ArrayList<>();
        min = new ArrayList<>();
    }
    
    public void push(int val) {
        stack.add(val);
        if (min.size() == 0 || val <= min.get(min.size() - 1)) {
            min.add(val);
        }
    }
    
    public void pop() {
        if (stack.size() > 0) {
            int val = stack.get(stack.size() - 1);
            stack.remove(stack.size() - 1);
            if (val == min.get(min.size() - 1)) {
                min.remove(min.size() - 1);
            }
        }
    }
    
    public int top() {
        return stack.get(stack.size() - 1);
    }
    
    public int getMin() {
        return min.get(min.size() - 1);
    }
}

// solution 2
class StackEntry {
    int min;
    int value;
    
    public StackEntry(int min, int value) {
        this.min = min;
        this.value = value;
    }
}

class MinStack {
    Stack<StackEntry> stack;
    
    public MinStack() {
        this.stack = new Stack();
    }
    
    public void push(int val) {
        if (stack.isEmpty()) {
            stack.add(new StackEntry(val, val));
        } else {
            int currMin = stack.peek().min < val ? stack.peek().min : val;
            stack.add(new StackEntry(currMin, val)); 
        }
    }
    
    public void pop() {
        if (!stack.isEmpty()) stack.pop();
    }
    
    public int top() {
        if (!stack.isEmpty()) {
            return stack.peek().value;
        }
        return -1;
    }
    
    public int getMin() {
        if (!stack.isEmpty()) {
            return stack.peek().min;
        }
        return -1;
    }
}

File: leetcode/easy/155-minStack.js
// Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

// push(x) -- Push element x onto stack.
// pop() -- Removes the element on top of the stack.
// top() -- Get the top element.
// getMin() -- Retrieve the minimum element in the stack.
 
// Example 1:

// Input
// ["MinStack","push","push","push","getMin","pop","top","getMin"]
// [[],[-2],[0],[-3],[],[],[],[]]

// Output
// [null,null,null,null,-3,null,0,-2]

// Explanation
// MinStack minStack = new MinStack();
// minStack.push(-2);
// minStack.push(0);
// minStack.push(-3);
// minStack.getMin(); // return -3
// minStack.pop();
// minStack.top();    // return 0
// minStack.getMin(); // return -2

// first pass solution
class MinStack {
    constructor() {
        this.storage = [];
    }
    
    push(num) {
        this.storage.push(num);
    }
    
    pop() {
        return this.storage.pop();
    }
    
    top() {
        return this.storage[this.storage.length - 1];
    }
    
    getMin() {
        return Math.min(...this.storage);
    } 
}
